elastalert/ruletypes.py

```diff
@@ -695,73 +696,123 @@ def __init__(self, rule, args=None):
+        self.es = kibana_adapter_client(self.rules)
+        window_size = datetime.timedelta(**self.rules.get('terms_window_size', {'days': 7}))
+        
+
         if args and hasattr(args, 'start') and args.start:
             end = ts_to_dt(args.start)
         elif 'start_date' in self.rules:
             end = ts_to_dt(self.rules['start_date'])
         else:
             end = ts_now()
+
         start = end - window_size
         step = datetime.timedelta(**self.rules.get('window_step_size', {'days': 1}))
-
+        
         for field in self.fields:
             tmp_start = start
             tmp_end = min(start + step, end)
-
-            time_filter = {self.rules['timestamp_field']: {'lt': self.rules['dt_to_ts'](tmp_end), 'gte': self.rules['dt_to_ts'](tmp_start)}}
-            query_template['filter'] = {'bool': {'must': [{'range': time_filter}]}}
-            query = {'aggs': {'filtered': query_template}, 'size': 0}
-
-            if 'filter' in self.rules:
-                for item in self.rules['filter']:
-                    query_template['filter']['bool']['must'].append(item)
-
-            # For composite keys, we will need to perform sub-aggregations
-            if type(field) == list:
-                self.seen_values.setdefault(tuple(field), [])
-                level = query_template['aggs']
-                # Iterate on each part of the composite key and add a sub aggs clause to the elastic search query
-                for i, sub_field in enumerate(field):
-                    if self.rules.get('use_keyword_postfix', True):
-                        level['values']['terms']['field'] = add_raw_postfix(sub_field, True)
-                    else:
-                        level['values']['terms']['field'] = sub_field
-                    if i < len(field) - 1:
-                        # If we have more fields after the current one, then set up the next nested structure
-                        level['values']['aggs'] = {'values': {'terms': copy.deepcopy(field_name)}}
-                        level = level['values']['aggs']
-            else:
-                self.seen_values.setdefault(field, [])
-                # For non-composite keys, only a single agg is needed
-                if self.rules.get('use_keyword_postfix', True):
-                    field_name['field'] = add_raw_postfix(field, True)
-                else:
-                    field_name['field'] = field
+            query = self.get_new_term_query(tmp_start,tmp_end,field)
 
             # Query the entire time range in small chunks
             while tmp_start < end:
-                if self.rules.get('use_strftime_index'):
-                    index = format_index(self.rules['index'], tmp_start, tmp_end)
-                else:
-                    index = self.rules['index']
-                res = self.es.search(body=query, index=index, doc_type='elastalert_status', ignore_unavailable=True, timeout='50s')
+                from elastalert import elastalert
```

Comment:
this is not good. Prepare query in another helper class and reference that file in both elastalert and ruletypes